### ISSUE TITLE
Alternative syntax

### ISSUE THREAD

**Issue by hax** [devlato/tc39-proposal-key-exclusion-syntax-in-object-spread#4]
Instead of `{-key, ...o}`:

1. `{...o, key: delete}`
2. `{...o, delete key}`
3. `{...o, key: void}`
4. `{...o, void key}`

Note, with such alternative syntax, I feel order could be preserved , so they occur after spread. (The only exception might be `void key` syntax which looks much declarative than others.)

For order independency, we could introduce a separate syntax, for example `!!` postfix.

- `{key!!: value, ...o}` (same as `{+key: value, ...o}` as current proposed syntax)
- `{void key!!, ...o}` (same as `{-key: value, ...o}` as current proposed syntax)

**Comment by devlato** [comment]
Thanks, @hax!

What is the use case for order dependence? I'm not sure whether order dependence would be worth implementing at all since it wouldn't give much of a benefit compared to what I proposed. Let me explain it case by case.

Let's start with a simple case where we have two objects (and mathematical induction will allow us to extrapolate it to any number of objects):
* object `a`:
  ```js
  const a = {
    foo: 10,
    bar: 'hey',
  };
  ```
* and object `b`:
  ```js
  const b = {
    fuzz: 10,
    buzz: 'hey',
  };
  ```

Then, imagine that we're using object spread to create a new object named `result`:
```js
const result = {
   ...a,
   ...b,
};
```

With the above example, we could apply key removal operator at one of the following places (no matter whether the operator would be order-dependent or order-independent):
* before `...a`
* after `...a` and before `...b`
* after `...b`

Let's dive deeper into every case, and compare order-dependent (here and below, `OD` abbreviation will be used) flow with the order-independent (let's use `OI` for that) one. Also, let's use one of the syntaxes you proposed (`delete key`), just in order to make the illustration easier.

_(Note: I assume for dynamic keys, you'd propose to use something like `{ delete [key] }`?)_

------------------

### Removing key before executing `...a`
In this case, we'd have the following code:
```js
const result = {
  delete key,
   ...a,
   ...b,
};
```

There are four sub-cases. Let's walk through them.
* `key` exists in `a` and does not exist in `b`:
  * `OD`: will make no difference, since it's applied before the created `result` object receives the keys and values from `a`. The `OD` operator here is useless – if we didn't use it, the `result` still would be the same
  * `OI`: will remove the key from the `result`, by design
* `key` exists in `b` and does not exist in `a`:
  * `OD`: will make no difference, since it's applied before the created `result` object receives the keys and values from `a`. The `OD` operator here is useless – if we didn't use it, the `result` still would be the same 
  * `OI`: will remove the key from the `result`, by design
* `key` exists in both `a` and `b`:
  * `OD`: will make no difference, since it's applied before the created `result` object receives the keys and values from `a`. The `OD` operator here is useless – if we didn't use it, the `result` still would be the same 
  * `OI`: will remove the key from the `result`, by design
* `key`'s present neither in `a` nor in `b`:
  * `OD`: will make no difference, since it's applied before the created `result` object receives the keys and values from `a`. The `OD` operator here is useless – if we didn't use it, the `result` still would be the same 
  * `OI`: will make no difference to `result`, since's neither in `a` nor `b` and thus, can't be in `result`

Conclusion: `OD` would make no difference to the `result`, while `OI` would serve as a guard and make sure that the key will be removed from the `result`

------------------

### Removing key after `...a` and before `...b`
In this case, we'd have the following code:
```js
const result = {
   ...a,
  delete key,
   ...b,
};
```

There are four sub-cases. Let's walk through them.
* `key` exists in `a` and does not exist in `b`:
  * `OD`: will remove the `key` from `a` and thus, from the `result` (since `b` does not have it)
  * `OI`: will remove the key from the `result`, by design
* `key` exists in `b` and does not exist in `a`:
  * `OD`: will leave the key in the result, the value will come from `b`. The `OD` operator here is useless – if we didn't use it, the `result` still would be the same 
  * `OI`: will remove the key from the `result`, by design
* `key` exists in both `a` and `b`:
  * `OD`: will leave the key in the result, the value will come from `b`. The `OD` operator here is useless – if we didn't use it, the `result` still would be the same 
  * `OI`: will remove the key from the `result`, by design
* `key`'s present neither in `a` nor in `b`:
  * `OD`: will make no difference, since it's applied before the created `result` object receives the keys and values from `a`. The `OD` operator here is useless – if we didn't use it, the `result` still would be the same
  * `OI`: will make no difference to `result`, since's neither in `a` nor `b` and thus, can't be in `result`

Conclusion: applying `OD` would be useless or would generate the same `result` as `OI`. At the same time, `OI` would serve as a guard and make sure that the key will be removed from the `result`

------------------

### Removing key after `...b`
In this case, we'd have the following code:
```js
const result = {
   ...a,
   ...b,
  delete key,
};
```

There are four sub-cases. Let's walk through them.
* `key` exists in `a` and does not exist in `b`:
  * `OD` will remove the key from the `result`
  * `OI`: will remove the key from the `result`, by design
* `key` exists in `b` and does not exist in `a`:
  * `OD` will remove the key from the `result`
  * `OI`: will remove the key from the `result`, by design
* `key` exists in both `a` and `b`:
  * `OD` will remove the key from the `result`
  * `OI`: will remove the key from the `result`, by design
* `key`'s present neither in `a` nor in `b`:
  * `OD` will make no difference to `result`, since's neither in `a` nor `b` and thus, can't be in `result`
  * `OI`: will make no difference to `result`, since's neither in `a` nor `b` and thus, can't be in `result`

Conclusion: `OD` would work in the same way as `OI`

------------------

From all the above, I don't see benefits of order-dependent implementation compared to order-independent :) Keys can be overridden by a following object in the spread and thus, don't need to be explicitly removed. The main use case is guarding the result of the spread against receiving the undesired (and specified with the proposed syntax) keys, and order-independent syntax serves as a better guarantee (and potentially, can be used better by typescript and linters, since it provides more type information and can be statically checked).

WDYT?

------------------

Regarding the syntax you proposed, I think there are a few cons:
1. `{...o, key: delete}` – `delete` is a reserved word by in some cases, can be a variable and can cause conflict and incompatibility (e.g. when created as a `window` property in a browser runtime).
2. `{...o, delete key}` – this would look a bit strange with dynamic keys. Would it look like `{...o, delete [key]}` when dynamic keys (stored in a variable) are used?
3. `{...o, key: void}` – `void` has another use case in JS (`void 0 === undefined`). Also, the same issues as in `(1)` exist
4. ` {...o, void key} – the same concerns as about `(2)`

**Comment by ljharb** [comment]
@devlato order dependence only matters when `a` and `b` have the _same key_. Map out your scenarios with that?

**Comment by devlato** [comment]
@ljharb could you speculate more on this, mate? I did illustrate this scenario above — unless you mean applying the order independent operator to a prepending argument in the object spread, rather than to the result that spread would produce at every step

**Comment by ljharb** [comment]
ah, sorry, i only looked at your simple object case (long posts are hard to parse)

If i do `{ …a, delete x, …b }` and x exists in b, the resulting object *must* have x.

**Comment by devlato** [comment]
@ljharb oh right, so is it the case where we'd want to always avoid receiving a value for the `key` from `a` but at the same time, keep it from `b` if `b` has it? Similarly to the following example:

```js
const a = { x: 10, y: 20 };
const b = { y: 40, z: 50 };
const c = { x: 20, y: 50, z: 30 };

const resultAB = { ...a, delete x, ...b }; // If `x` is not in `b`, we don't want it to be in the final result either
const resultAC = { ...a, delete x, ...c }; // However, if `x` is in `c`, we want to keep it
```

**Comment by ljharb** [comment]
Yes, precisely that. That’s why order matters in object spread now, and why it must matter in any future syntax proposal that expands object spread.

**Comment by hax** [comment]
I just wrote the possible alternative syntax, and I'm neutral on whether it's better than current proposed syntax or not. Actually I also feel order independency is a much interesting use case, more declarative. I just try to explore the syntax spaces and see what other possibilities are.

**Comment by devlato** [comment]
@hax @ljharb thanks for your valuable inputs, gentlemen. Btw I'm looking for a champion to help move this proposal forward – would any of you kindly help with that? 🙂

**Comment by devlato** [comment]
Thanks, folks—the issue has been addressed in [the new edition of the proposal](https://github.com/devlato/tc39-proposal-key-exclusion-syntax-in-object-spread).

**Comment by hax** [comment]
Note: [`void` discard binding proposal](https://github.com/tc39/proposal-discard-binding?tab=readme-ov-file#object-binding-and-assignment-patterns) already include `{...o, key: void}` syntax. There is also discussion of using `-` as dicard token: https://github.com/tc39/proposal-discard-binding/issues/2

**Comment by devlato** [comment]
Thanks @hax. From my understanding, the discard binding proposal only suggests the `void` syntax for destructuring and not for object spread—is that correct?

cc @rbuckton from [tc39/proposal-discard-binding](https://github.com/tc39/proposal-discard-binding)
